Instadapp integration: add Liquity connector

edmulraney opened the pull request:
Adds support for Liquity to the Instadapp connectors repo.

To run Instadapp's tests you need an Alchemy API key (see the example.env file), and a dummy private key (not used in any of the Liquity tests but is required to build).

Todo:
- Remove `.only()` from Liquity tests, deliberately leaving this in for now so its easier to review this PR

edmulraney committed:
add liquity connector

bingen commented on contracts/mainnet/connectors/liquity/events.sol:
```diff
@@ -0,0 +1,42 @@
+pragma solidity ^0.7.0;
```
Is this for consistency with Instadapp contracts? Trail of Bits recommends 0.7.5 - 0.7.6

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+    }
+
+    constructor() {
+        console.log("Liquity Connector contract deployed at", address(this));
```
I  guess we should comment this out before real deployment.

edmulraney replied:
https://github.com/liquity/dsa-connectors/commit/97db94fe595330743385f26c30e15ab28cc18063

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+
+        // Allow other spells to use the borrowed amount
+        setUint(setId, borrowAmount);
+        _eventName = "LogOpen(address,uint,uint,uint,uint,uint)";
```
Not sure what this is used for, but if it’s later hashed with `keccak256`, `uint` should probably be `uint256`.

edmulraney replied:
Aren't they the same?

edmulraney replied:
Aren't they the same?

bingen replied:
Yes, but as the hash is done just from text, the result is different. You can try it here:
https://emn178.github.io/online-tools/keccak_256.html
Try with `LogOpen(address,uint,uint,uint,uint,uint)` and with `LogOpen(address,uint256,uint256,uint256,uint256,uint256)` and you will see the result is different.

edmulraney replied:
https://github.com/liquity/dsa-connectors/commit/97db94fe595330743385f26c30e15ab28cc18063

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+        uint getRepayId,
+        uint setBorrowId
+    ) external payable returns (string memory _eventName, bytes memory _eventParam) {
+        AdjustTrove memory adjustTrove;
```
Is this to avoid “stack too deep” issues?

edmulraney replied:
Yeah only added it to avoid that issue

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+        uint setId
+    ) external payable returns (string memory _eventName, bytes memory _eventParam) {
+        // User can either send ETH directly or have it collected from a previous spell
+        depositAmount = getUint(getId, depositAmount);
```
If I understand this correctly, if `getId` is not zero, it would override `depositAmount` passed in value. Would it make sense to check that user only passes one of them as non-zero, to prevent user mistakes?

edmulraney replied:
https://github.com/liquity/dsa-connectors/pull/1/commits/97db94fe595330743385f26c30e15ab28cc18063

bingen reviewed (commented):
I did a first superficial pass, as I still need to understand well all the Instadapp stuff, but it looks like impressive work!

edmulraney commented on test/liquity/liquity.test.js:
```diff
@@ -0,0 +1,1133 @@
+// Instadapp uses a fake address to represent native ETH
+const { eth_addr: ETH_ADDRESS } = require("../../scripts/constant/constant");
+
+describe.only("Liquity", () => {
```
Deliberately leaving `.only` until PR review is finished

edmulraney committed:
add recommended solidity version

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+        setUint(setWithdrawId, withdrawAmount);
+
+        // Allow other spells to use the borrowed amount
+        setUint(setBorrowId, borrowAmount);
```
I guess the same “TODO” comment about the fee applies here too.

edmulraney replied:
Same as this: https://github.com/liquity/dsa-connectors/pull/1/commits/97db94fe595330743385f26c30e15ab28cc18063#r645495186

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+        );
+
+        // Allow other spells to use the borrowed amount
+        setUint(setId, borrowAmount);
```
I guess the same “TODO” comment about the fee applies here.

edmulraney replied:
Borrow amount is the exact amount they will borrow, not the total debt, so I don't believe there's anything to change here

bingen replied:
Yes, you are right, but so it’s for `withdrawLUSD`, right?

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+     * @dev Withdraw remaining ETH balance from user's redeemed Trove to their DSA
+     * @notice Claim remaining collateral from Trove
+    */
+    function claimCollateralFromRedemption() external returns(string memory _eventName, bytes memory _eventParam) {
```
Shouldn’t we have here a `setUint` for the collateral obtained?

edmulraney replied:
https://github.com/liquity/dsa-connectors/commit/97db94fe595330743385f26c30e15ab28cc18063#r645495186

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+        amount = getUint(getId, amount);
+
+        stabilityPool.provideToSP(amount, frontendTag);
```
In all operations it would be nice  to have `setUint`’s for LQTY and ETH gains.

edmulraney replied:
https://github.com/liquity/dsa-connectors/pull/1/commits/97db94fe595330743385f26c30e15ab28cc18063

bingen commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,347 @@
+        uint getId
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+        amount = getUint(getId, amount);
+        staking.stake(amount);
```
Same here, in all staking operations it would be nice to have setUint’s for LUSD and ETH gains.

edmulraney replied:
https://github.com/liquity/dsa-connectors/pull/1/commits/97db94fe595330743385f26c30e15ab28cc18063

edmulraney replied:
https://github.com/liquity/dsa-connectors/pull/1/commits/5be5cb68668172047e29323092c574d4155f80f8

bingen reviewed (approved):
I left some more minor comments below, but looking good! I have some generic questions too:

- I guess it’s documented somewhere, but how is authorization for the different actions managed?
- I see there has been a recent upgrade from “DSA” to “DSL”, but besides tokenization and upgradability, I don’t know the exact technical differences yet. Is this targeting the latter?
- I see that `getUint` cleans the storage slot, which is nice to save gas. But what happens if a `setUint` is not consumed? Is there an easy/standard way to delete it?

edmulraney committed:
add test for instadapp event return values. address PR comments:
    - add setUints to all applicable functions
    - use uint256 rather than uint for event data

edmulraney committed:
include all gains from staking/stability-pool when performing operations on them

RickGriff commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,426 @@
+        address upperHint,
+        address lowerHint,
+        uint getId,
+        uint setId
```
So these `getId`  / `setId` variables are used to specify whether the the function should use ETH from the previous spell / store the LUSD amount used here for the next spell.

Who chooses those id values, and what prevents "wrong" values being used? I guess this is handled by the InstaDapp infrastructure?

edmulraney replied:
They are arbitrary values, as long as the values are unique you can guarantee your spell will work. It overwrites any existing amount stored at the ID. The dev just chooses IDs that are unique to their spell

RickGriff commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,426 @@
+     * @param depositAmount The amount of ETH to deposit
+     * @param maxFeePercentage The maximum borrow fee that this transaction should permit 
+     * @param borrowAmount The amount of LUSD to borrow
+     * @param upperHint Address of the Trove near the upper bound of where the user's Trove will now sit in the ordered Trove list
```
minor nit: this is the ideal situation, but it could happen that the list shifts between tx broadcast and mining, such that the hints are out of date and the system needs to step beyond the upper or lower hint to get the right spot.  Or if the wrong hints were passed anyway.

So I guess we should say they are estimates of bounds - or just "where the user's Trove _should_ now sit"

RickGriff commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,426 @@
+        address lowerHint,
+        uint getId
+    ) external payable returns (string memory _eventName, bytes memory _eventParam)  {
+        if (getId != 0 && amount != 0) {
```
These reverts make good sense!

RickGriff commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,426 @@
+    ) external payable returns (string memory _eventName, bytes memory _eventParam)  {
+        borrowerOperations.withdrawLUSD(maxFeePercentage, amount, upperHint, lowerHint);
+
+        setUint(setId, amount); // TODO: apply fee / get exact amount borrowed (with the fee applied)
```
Is this still TODO or is it OK to store the drawn LUSD amount?

edmulraney replied:
Yeah the drawn LUSD amount is the value we want to share with future spells

RickGriff commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,426 @@
+     * @param frontendTag Address of the frontend to make this deposit against (determines the kickback rate of rewards)
+     * @param getDepositId Optional storage slot to retrieve the LUSD from
+     * @param setEthGainId Optional storage slot to store any ETH gains in
+     * @param setLqtyGainId Optional storage slot to store any ETH gains in
```
Looks like a typo here, I suppose it should be "LQTY gains"

RickGriff commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,426 @@
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+        amount = getUint(getDepositId, amount);
+        uint ethGain = stabilityPool.getDepositorETHGain(address(this));
+        uint lqtyGain = stabilityPool.getDepositorLQTYGain(address(this));
```
Because of the way LQTY rewards work, the returned value from `getDepositorLQTYGain` will be usually be slightly out of date, and slightly less than the actual LQTY paid out when the depositor does `provideToSP`.

How out of date it is depends on how frequently `StabilityPool` actions occur (by any depositor), but there's usually a 10-30 minute lag between them, and so `lqtyGain` will be a little low.

Not sure how much it matters here if that stored value is slightly less than the actual LQTY paid out.

RickGriff commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,426 @@
+     * @param amount Amount of LUSD to withdraw from Stability Pool
+     * @param setWithdrawId Optional storage slot to store the withdrawn LUSD
+     * @param setEthGainId Optional storage slot to store any ETH gains in
+     * @param setLqtyGainId Optional storage slot to store any ETH gains in
```
Should be "LQTY gains"

RickGriff commented on contracts/mainnet/connectors/liquity/main.sol:
```diff
@@ -0,0 +1,426 @@
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+        stabilityPool.withdrawFromSP(amount);
+        uint ethGain = stabilityPool.getDepositorETHGain(address(this));
+        uint lqtyGain = stabilityPool.getDepositorLQTYGain(address(this));
```
Again `getDepositorLQTYGain` returns slightly less than the LQTY that will be paid out

RickGriff commented on test/liquity/liquity.test.js:
```diff
@@ -0,0 +1,1987 @@
+      });
+    });
+  });
+});
```
These looks like really great tests!

RickGriff reviewed (commented):
Awesome! Overall it looks great - just a couple of questions, and a point about the impact of `getDepositorLQTYGain`, which tends to return slightly out-of-date values.

edmulraney committed:
add tests to ensure getId/setIds all work

edmulraney committed:
log the eth and lqty gains from stability pool operations. general tidy up
